internal/message/watchutil.go

```diff
@@ -20,20 +45,36 @@ type Update[K comparable, V any] watchable.Update[K, V]
 // it handles the case where the watchable.Map already contains
 // entries before .Subscribe is called.
 func HandleSubscription[K comparable, V any](
+	meta Metadata,
 	subscription <-chan watchable.Snapshot[K, V],
-	handle func(Update[K, V]),
+	handle func(updateFunc Update[K, V], errChans chan error),
 ) {
+	errChans := make(chan error, 10)
+	go func() {
+		for err := range errChans {
+			logger.WithValues("runner", meta.Runner).Error(err, "observed an error")
+			messageSubscribedErrorsTotal.With(meta.LabelValues()...).Increment()
+		}
+	}()
+
 	if snapshot, ok := <-subscription; ok {
 		for k, v := range snapshot.State {
+			startHandleTime := time.Now()
 			handle(Update[K, V]{
 				Key:   k,
 				Value: v,
-			})
+			}, errChans)
+			messageSubscribedTotal.With(meta.LabelValues()...).Increment()
+			messageSubscribedDurationSeconds.With(meta.LabelValues()...).Record(time.Since(startHandleTime).Seconds())
 		}
 	}
 	for snapshot := range subscription {
+		messageDepth.With(meta.LabelValues()...).Record(float64(len(subscription)))
 		for _, update := range snapshot.Updates {
-			handle(Update[K, V](update))
+			startHandleTime := time.Now()
+			handle(Update[K, V](update), errChans)
+			messageSubscribedTotal.With(meta.LabelValues()...).Increment()
+			messageSubscribedDurationSeconds.With(meta.LabelValues()...).Record(time.Since(startHandleTime).Seconds())
```

Comment:
make these lines into a function will make sure you will not forget update anyone of those metrics.